### PR TITLE
Fix traffic congestion rendering issues.

### LINKED ISSUE #2423: Investigate route gradient stop property declaration in NavigationMapView
Following up on https://github.com/mapbox/mapbox-navigation-ios/pull/2377#discussion_r445168323,

`NavigationMapView` declares two gradient stop dictionaries in the form of a tuple, used to calculate the traffic line gradients for both the main route line and its casing:

```swift
typealias RouteGradientStops = (line: [CGFloat: UIColor], casing: [CGFloat: UIColor])
private var routeGradientStops = RouteGradientStops(line: [:], casing: [:])
```

It was suggested in https://github.com/mapbox/mapbox-navigation-ios/pull/2377#discussion_r445168323 that this be changed to two parallel properties instead, such as:

```swift
private var routeGradientStops: [CGFloat: UIColor] = [:]
private var routeGradientCasingStops: [CGFloat: UIColor] = [:]
```

However, this declaration causes the gradient stops to be miscalculated - we should look into why this is happening.

| Screen Shot 2020-06-24 at 3 58 47 PM | Screen Shot 2020-06-24 at 3 57 33 PM |
|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Tuple                                                                                                                                                                   | Two parallel properties                                                                                                                                                 |

\* note - the line caps are not rounded in these two screenshots, but since those were taken that has been corrected.

cc @1ec5

### PR DESCRIPTION
**Pull request by MaximAlien**
Closing #2574.

I've slightly re-worked original implementation. Now there are such methods which allow to add route layers:
```swift
addMainRouteLayer(_:source:lineGradient:)
addMainRouteCasingLayer(_:source:lineGradient:below:)
addAlternateRoutesLayer(_:source:below:)
addAlternateRoutesCasingLayer(_:source:below:)
```

Such methods are used to create line gradient:
```swift
routeLineGradient(_:fractionTraveled:)
routeCasingGradient(_:)
```

In case if route layer was already added route line gradient with congestion segments will be retrieved out of it.
For route casing color is always the same (without congestions), but fading is also added depending on route progress.

Current PR also fixes #2423 and #2512.

### CODE REVIEW

File: MapboxNavigation/NavigationMapView.swift [hunk 1]
```diff
@@ -487,126 +477,202 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
...
+        style.insertLayer(mainRouteCasingLayer, below: layer)
+        
         return mainRouteCasingLayer
     }
-
-    func alternativeRouteStyleLayer(identifier: String, source: MGLSource) -> MGLLineStyleLayer {
-        let alternateRoutesLayer = MGLLineStyleLayer(identifier: identifier, source: source)
+    
+    @discardableResult func addAlternateRoutesLayer(_ style: MGLStyle, source: MGLSource, below layer: MGLStyleLayer) -> MGLStyleLayer {
+        let customAlternateRoutesLayer = navigationMapViewDelegate?.navigationMapView(self,
```
Comment (MaximAlien) [inline MapboxNavigation/NavigationMapView.swift:612]: At this point with use `NavigationMapViewDelegate.navigationMapView(_:alternativeRouteStyleLayerWithIdentifier:source:)` and `NavigationMapViewDelegate.navigationMapView(_:alternateRouteCasingStyleLayerWithIdentifier:source:)`. @1ec5, maybe it'd be better to use the same convention here? As I can see `alternate` is mostly used in `NavigationMapView`, maybe we should use the same naming in public API method?

File: MapboxNavigation/NavigationMapView.swift [hunk 2]
```diff
@@ -487,126 +477,202 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
...
+                                                                                      alternativeRouteStyleLayerWithIdentifier: StyleLayerIdentifier.alternateRoutes,
```
Comment (MaximAlien) [inline MapboxNavigation/NavigationMapView.swift:613]: At this point we use `NavigationMapViewDelegate.navigationMapView(_:alternativeRouteStyleLayerWithIdentifier:source:)` and `NavigationMapViewDelegate.navigationMapView(_:alternateRouteCasingStyleLayerWithIdentifier:source:)`. @1ec5, maybe it'd be better to use the same convention here? As I can see `alternate` is mostly used in `NavigationMapView`, maybe we should use the same naming in public API method?

Comment (1ec5) [reply MapboxNavigation/NavigationMapView.swift:613]: It should be “alternative” consistently. “Alternate” is easily read as a verb in this context, and it has a slightly different meaning anyways (alternating back and forth, not the alternative to something).

File: MapboxNavigation/NavigationMapView.swift [hunk 3]
```diff
@@ -487,126 +477,202 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
...
+        
         self.routes = routes
+        
+        let allRoutesShape = navigationMapViewDelegate?.navigationMapView(self, shapeFor: routes) ?? shape(for: routes, legIndex: legIndex)
+        let allRoutesSource = addAllRoutesSource(style, shape: allRoutesShape)
+        
+        let mainRouteCasingShape = navigationMapViewDelegate?.navigationMapView(self, shapeFor: [mainRoute]) ?? shape(for: [mainRoute], legIndex: legIndex)
+        let mainRouteCasingSource = addMainRouteCasingSource(style, shape: mainRouteCasingShape)
+        
+        let mainRouteLayer = addMainRouteLayer(style, source: allRoutesSource, lineGradient: routeLineGradient(mainRoute, fractionTraveled: 0.0))
```
Comment (1ec5) [inline MapboxNavigation/NavigationMapView.swift:491]: The new helper methods make this method much easier to follow, thanks!

**Pull request merged**